### PR TITLE
[jamf_pro] Various minor improvements and fixes

### PR DESCRIPTION
**Pull request by chrisberkhout**
## Proposed commit message

```
[jamf_pro] Various minor improvements and fixes

Mostly tidying of the README and pipelines, but also some small fixes.
```

Reviews should probably focus on the pipelines.

## Checklist

- [ ] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [ ] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [ ] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).

## Related issues

- Relates #10470
- Relates #10026

### CODE REVIEW

File: packages/jamf_pro/manifest.yml [hunk 1]
```diff
@@ -1,7 +1,7 @@
 format_version: 3.1.5
 name: jamf_pro
 title: "Jamf Pro"
-version: 0.1.0
+version: 0.1.1
 source:
   license: "Elastic-2.0"
 description: "A connector for Jamf Pro API"
```
Comment (jamiehynds) [inline packages/jamf_pro/manifest.yml:7]: @chrisberkhout minor nit, but could we also adjust this description inline with other integrations. 

"Collect logs and inventory data from Jamf Pro with Elastic Agent"

File: packages/jamf_pro/data_stream/events/elasticsearch/ingest_pipeline/default.yml [hunk 2]
```diff
@@ -13,12 +14,14 @@ processors:
 - convert:
     field: jamf_pro.events.event.management_id
     type: string
-    if: "ctx.jamf_pro.events.event?.management_id != null && !(ctx.jamf_pro.events.event?.management_id  instanceof String)"
+    if: |
```
Comment (efd6) [inline packages/jamf_pro/data_stream/events/elasticsearch/ingest_pipeline/default.yml:17]:
```suggestion
    if: >
```

File: packages/jamf_pro/data_stream/events/elasticsearch/ingest_pipeline/default.yml [hunk 3]
```diff
@@ -13,12 +14,14 @@ processors:
...
+      ctx.jamf_pro.events.event?.management_id != null &&
+      !(ctx.jamf_pro.events.event.management_id instanceof String)
 
 - rename:
     field: jamf_pro.events.event.computer
     target_field: jamf_pro.events.event.is_computer
-    if: (ctx.jamf_pro.events.event.computer === true || ctx.jamf_pro.events.event.computer === false)
+    if: "[true, false].contains(ctx.jamf_pro.events.event.computer)"
```
Comment (efd6) [inline packages/jamf_pro/data_stream/events/elasticsearch/ingest_pipeline/default.yml:24]:
```suggestion
    if: ctx.jamf_pro?.events?.event?.computer instanceof boolean 
```

File: packages/jamf_pro/data_stream/events/elasticsearch/ingest_pipeline/default.yml [hunk 4]
```diff
@@ -67,44 +70,47 @@ processors:
 - date:
     field: jamf_pro.events.event.last_update
     target_field: jamf_pro.events.event.last_update
-    if: "ctx.jamf_pro.events?.last_update != null"
+    if: "ctx.jamf_pro.events.event.last_update != null"
```
Comment (efd6) [inline packages/jamf_pro/data_stream/events/elasticsearch/ingest_pipeline/default.yml:73]:
```suggestion
    if: ctx.jamf_pro?.events?.event?.last_update != null
```
Unless we're confident in the existence of the `ctx.jamf_pro.events.event` path. If we are, we should probably have a comment at the point from which we guarantee it exists. (similar below)

File: packages/jamf_pro/data_stream/events/elasticsearch/ingest_pipeline/default.yml [hunk 5]
```diff
@@ -67,44 +70,47 @@ processors:
...
-    if: "ctx.jamf_pro.events.event?.ip_address != null && ctx.jamf_pro.events.event?.ip_address != ''"
+    if: "ctx.jamf_pro.events.event.ip_address != null && ctx.jamf_pro.events.event.ip_address != ''"
     field: host.ip
     value: 
      - '{{{jamf_pro.events.event.ip_address}}}'
     ignore_empty_value: true
 
 - set:
-    if: "ctx.jamf_pro.events.event?.computer != null && ctx.jamf_pro.events.event?.computer?.reported_ip_address != null && ctx.jamf_pro.events.event?.computer?.reported_ip_address != ''"
+    if: |
```
Comment (efd6) [inline packages/jamf_pro/data_stream/events/elasticsearch/ingest_pipeline/default.yml:89]:
```suggestion
    if: >
```

File: packages/jamf_pro/data_stream/events/elasticsearch/ingest_pipeline/default.yml [hunk 6]
```diff
@@ -137,20 +143,28 @@ processors:
...
     target_field: host.geo
     ignore_missing: true
 
-- set:
+- append:
     field: related.user
-    copy_from: user.name
-    ignore_empty_value: true
+    value: "{{{user.name}}}"
+    if: ctx.user?.name != null && ctx.user?.name != ""
```
Comment (efd6) [inline packages/jamf_pro/data_stream/events/elasticsearch/ingest_pipeline/default.yml:154]:
```suggestion
    if: ctx.user?.name != null && ctx.user.name != ""
```

File: packages/jamf_pro/data_stream/events/elasticsearch/ingest_pipeline/default.yml [hunk 7]
```diff
@@ -137,20 +143,28 @@ processors:
     ignore_empty_value: true
 
 - geoip:
-    if: ctx.host?.ip != null && ctx.host?.ip != '' 
+    if: ctx.host?.ip != null && ctx.host?.ip != ''
```
Comment (efd6) [inline packages/jamf_pro/data_stream/events/elasticsearch/ingest_pipeline/default.yml:146]:
```suggestion
    if: ctx.host?.ip != null && ctx.host.ip != ''
```

File: packages/jamf_pro/data_stream/events/elasticsearch/ingest_pipeline/default.yml [hunk 8]
```diff
@@ -137,20 +143,28 @@ processors:
...
     field: related.user
-    copy_from: user.name
-    ignore_empty_value: true
+    value: "{{{user.name}}}"
+    if: ctx.user?.name != null && ctx.user?.name != ""
+
+- append:
+    field: related.user
+    value: "{{{user.email}}}"
+    if: ctx.user?.email != null && ctx.user?.email != ""
```
Comment (efd6) [inline packages/jamf_pro/data_stream/events/elasticsearch/ingest_pipeline/default.yml:159]:
```suggestion
    if: ctx.user?.email != null && ctx.user.email != ""
```

File: packages/jamf_pro/data_stream/inventory/agent/stream/cel.yml.hbs [hunk 9]
```diff
@@ -30,119 +30,128 @@ publisher_pipeline.disable_host: true
...
+  	}.format_query()
+  ).with(
+  	{
+  		"Header": {
+  			"Content-Type": ["application/json"],
+  		},
+  	}
+  ).do_request().as(resp,
+  	bytes(resp.Body).decode_json().as(body,
+  		(resp.StatusCode != 200) ?
```
Comment (efd6) [inline packages/jamf_pro/data_stream/inventory/agent/stream/cel.yml.hbs:53]: If these are on the same line, celfmt will respect that and reduce the indentation below.

File: packages/jamf_pro/data_stream/events/elasticsearch/ingest_pipeline/default.yml (hunk 2 above)
Comment (chrisberkhout) [reply packages/jamf_pro/data_stream/events/elasticsearch/ingest_pipeline/default.yml:17]: Done.

File: packages/jamf_pro/data_stream/events/elasticsearch/ingest_pipeline/default.yml (hunk 3 above)
Comment (chrisberkhout) [reply packages/jamf_pro/data_stream/events/elasticsearch/ingest_pipeline/default.yml:24]: Did this and the next one like this:
```diff
diff --git a/packages/jamf_pro/data_stream/events/elasticsearch/ingest_pipeline/default.yml b/packages/jamf_pro/data_stream/events/elasticsearch/ingest_pipeline/default.yml
index 5fc1d4d20e..6bef2c6ada 100644
--- a/packages/jamf_pro/data_stream/events/elasticsearch/ingest_pipeline/default.yml
+++ b/packages/jamf_pro/data_stream/events/elasticsearch/ingest_pipeline/default.yml
@@ -9,7 +9,7 @@ processors:
 - rename:
     field: json
     target_field: jamf_pro.events
-    if: ctx.json != null
+    # all valid events will now have jamf_pro.events.{event,webhook}
 
 - convert:
     field: jamf_pro.events.event.management_id
@@ -21,7 +21,7 @@ processors:
 - rename:
     field: jamf_pro.events.event.computer
     target_field: jamf_pro.events.event.is_computer
-    if: "[true, false].contains(ctx.jamf_pro.events.event.computer)"
+    if: ctx.jamf_pro.events.event.computer instanceof boolean
 
 - script:
     description: Convert Additional Info keys to snake case.
@@ -70,7 +70,7 @@ processors:
 - date:
     field: jamf_pro.events.event.last_update
     target_field: jamf_pro.events.event.last_update
-    if: "ctx.jamf_pro.events.event.last_update != null"
+    if: ctx.jamf_pro.events.event.last_update != null
     formats:
       - UNIX_MS
```

File: packages/jamf_pro/data_stream/events/elasticsearch/ingest_pipeline/default.yml (hunk 4 above)
Comment (chrisberkhout) [reply packages/jamf_pro/data_stream/events/elasticsearch/ingest_pipeline/default.yml:73]: Yes, we're confident. Did it as above.

File: packages/jamf_pro/data_stream/events/elasticsearch/ingest_pipeline/default.yml (hunk 5 above)
Comment (chrisberkhout) [reply packages/jamf_pro/data_stream/events/elasticsearch/ingest_pipeline/default.yml:89]: Done.

File: packages/jamf_pro/data_stream/events/elasticsearch/ingest_pipeline/default.yml (hunk 8 above)
Comment (chrisberkhout) [reply packages/jamf_pro/data_stream/events/elasticsearch/ingest_pipeline/default.yml:159]: Done.

File: packages/jamf_pro/data_stream/events/elasticsearch/ingest_pipeline/default.yml (hunk 7 above)
Comment (chrisberkhout) [reply packages/jamf_pro/data_stream/events/elasticsearch/ingest_pipeline/default.yml:146]: Done.

File: packages/jamf_pro/data_stream/inventory/agent/stream/cel.yml.hbs (hunk 9 above)
Comment (chrisberkhout) [reply packages/jamf_pro/data_stream/inventory/agent/stream/cel.yml.hbs:53]: Nice. Done.

**Review by efd6 — APPROVED** [review]

**Comment by chrisberkhout** [comment]
Here is a summary of the main things that came up in this PR, which will help with future integration development.

#### Aligning input config, pipeline test input and pipeline logic:

- There was some [config missing from http_endpoint.yml.hbs](https://github.com/elastic/integrations/pull/11065/files#diff-d1459335820f7ab62af55712473e6023334b033ba43ca4aa27f4155772b9bab3) which meant that the `rename` at the start of the events pipeline wasn't being run, because the `json` key was never populated. It's tricky but very important to make sure that the pipeline test input data and the live events arrive at the pipeline in the same format.

- There was a related data structure/pipeline start error in the [inventory pipeline](https://github.com/elastic/integrations/pull/11065/files#diff-94f6a51477b87ffaea342ff01e1ef82131d2cb7945c71494fea8eb303ec3df2b) where fingerprinting was attempted on fields in root that were actually nested in "message".

#### Painless and the null-safe operator

- The null-safe operator, and Painless in general, can only be used if a processor option allows it. It's okay for `if` options, but values for `field` or `copy_from` options should be plain names, not Painless fragments. If you need to account for a name not existing, you can use options like `ignore_missing` or add an `if` condition so the processor only runs under the correct conditions.

- Use the null-safe operator after a thing that may or may not exist (not before). It's okay to use one key that doesn't exist (that will return null). So, in the example `ctx.jamf_pro.events.event?.ip_address != null`, if `event` exists the null-safe operator is redundant.

#### Append to make arrays

The `related.*` ECS fields should be arrays regardless of how many items they have, so use `append` rather than `set`.

#### CEL formatting

There is a new tool for formatting CEL programs: https://github.com/efd6/celfmt We encourage using this now. I recommend using it via [this bash alias](https://gist.github.com/chrisberkhout/e9d5dd0b00d738cc66ecebe9bda3e617#file-helpers-sh-L19).

#### README

I tidied up the README a bit. I added a little bit of text but mostly it was about cleaning up the formatting. Below there is an image showing the rendered versions side-by-side. The red arrows mark some of the more significant formatting improvements. The goal is to use structure, emphasis, etc. consistently to make it easier to read. Ideally the unrendered markdown will also be nicely formatted and readable.

readme-side-by-side-marked-up

**Pull request merged**